### PR TITLE
FEATURE: Read Examples table data from CSV file

### PR DESCRIPTION
**Pull request by rajanmanwansh**
This pull request introduces a new feature that allows users to dynamically populate Scenario Outline examples from a CSV file in Behave. This enhancement enables data-driven testing by simplifying the process of loading test data, making it easier for users to manage and update test cases with external data files.

Key Changes

1. Introduced a new feature to dynamically populate examples tables in Scenario Outline from a specified CSV file.
2. Added unit tests and functional tests to ensure the new feature works as expected.
3. Verified that the new changes do not affect the existing functionality.

```
**Usage**:

Feature: Example Feature

Scenario Outline: Example Scenario
    Given get <username> and <email> and <password>

    @from_file=testdata.csv
    Examples:
      | username | email   | password |
      | .                 | .           | .                |

    @from_file=testdata2.csv
    Examples:
      | username | email   | password |
      | .                 | .           | .                |
```

### CODE REVIEW

**Comment by jenisys** [comment]
BETTER, BUT:

* None of your files directly affects `behave`
* You can easily provide your solution as `behave extension` and use it or provide it for others
* The `features/environment.py` changes are helpful to understand how this functionality is provided (as: proof-of-concept). But because no feature actually uses it, it can be removed.

NOTES: Related to `environment.py`

* Provide the parts/section as functions to make them easier to use in other places and for other people (hint: REUSABLE CODE).
* One possibility can be to provide it as repository under: https://github.com/behave-contrib/ (if you are willing to maintain it and publish the package). Let me know, if that is an option for you (and I need the name of the repo).

**Comment by jenisys** [comment]
ALTERNATIVE:
* Provide the core functionality under `behave/contrib/csv_table_from_file.py`
* SEE: https://github.com/behave/behave/tree/main/behave/contrib

**Comment by rajanmanwansh** [comment]
@jenisys Please review the latest changes.

**Comment by jenisys** [comment]
OPEN ISSUES:
* [ ] `environment.py`: Move `else-body`, that provides the preprocessing and table-reading in `before_feature()` hook as function into `cvs_table_from_file.py`
* [ ] `cvs_table_from_file.py`: Needs a docstring that describes what is the purpose of this module and how should I use it. Basically the functionality that is currently in `environment.py`
* [ ] Any changes in `environment.py` are not needed. Functionality is currently not used there. It is now described in the docstring of the the new module.
* [ ] tests: Do not import from `features/environment.py`, use `behave.contrib.csv_table_from_file`
* [ ] Test module name: Should reflect the module your are testing, in this case `behave.contrib.cvs_table_from_file` like: `tests/unit/test_csv_table_from_file.py` or `tests/unit/contrib/test_csv_table_from_file.py`

LIMITATION:
* The functionality will currently cause problems if you use `FileLocation` on the command-line that refers to the table row.
* You should probably add this information to the docstring in the new module.

**Comment by rajanmanwansh** [comment]
Sure, will provide the fixes soon.

**Comment by jenisys** [comment]
Mmh:

* The name of the module is not the name of the function that you provide. Therefore, why change it (from the good and smaller module name that I provided and that you already used) ?!?
* Bail out early keeps indentation levels smaller and your code more readable:

```py
# BAD-CASE:
def read_examples_table_data_from_csv(example, file_path):
    if os.path.exists(file_path):
        …
    else:
        raise FileNotFoundError(f'no file found {file_path}')
```

```py
# GOOD-CASE:
def read_examples_table_data_from_csv(example, file_path):
    if not os.path.exists(file_path):
        raise FileNotFoundError(f'no file found {file_path}')
    
    # — NORMAL CASE:
    …
```

```py
# — BAD CASE:
def preprocess_and_read_examples_table_data_from_csv(…):
    for scenario in feature.walk_scenarios():  # HINT: Needs with_outlines=True
        if isinstance(scenario, ScenarioOutline):
            # — NORMAL CODE: …
```

```py
# — GOOD CASE:
def preprocess_and_read_examples_table_data_from_csv(…):
    for scenario_outline in feature.iter_scenario_outlines():
        # — NORMAL CODE: …
```

**Comment by jenisys** [comment]
Looks better, BUT:

* The module documentation is still missing (and my example was seemingly ignored)
* By looking over your implementation again, your implementation does not what is described in the docstring
* With my guiding hints, you need to retrieve the CVS filename from the `@from_file={csv_filename}` tag of the `Examples` table (if it has one). Only then, your functionality should be applied (meaning: Reading the table data from the `cvs_filename` and use it for the `Examples` table.
* The current implementation seems to use the `Feature` filename and replace its file-extension with `.csv`. If this cvs-file exists, the CVS table data is used on any `Examples.table` of a `ScenarioOutline` even if the `Examples` table is not marked with the `@from_file=…` tag (which is highly undesirable)
* The tests also do not cover this, just that some CVS filename is provided and loaded

NOT COVERED:
* What happens if the provided `filename` is not a CVS file ?!? (Missing in: implementation, tests)

**Comment by rajanmanwansh** [comment]
@jenisys  plz review the latest commit.

File: behave/contrib/csv_table_from_file.py [hunk 1]
```diff
@@ -53,25 +93,43 @@ def read_examples_table_data_from_csv(example, file_path):
...
-            continue
-
         for example in scenario_outline.examples:
-            read_examples_table_data_from_csv(example, csv_file_path)
-
+            csv_file_path = None
+
+            # Find the @from_file={csv_filename} tag
+            for tag in example.tags:
+                if tag.startswith('@from_file='):
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:118]: Tags from `behave` model elements (`Feature`, `Scenario`, `Examples`, etc.) have the `@` prefix stripped (after parsing). Therefore, comparing a tag with starts-with `@from_file=` will never find such a tag.

REPLACE-WITH: `tag.startswith("from_file=")`
NOTE:
* Indicates that you probably have not tested the core functionality.

File: behave/contrib/csv_table_from_file.py [hunk 2]
```diff
@@ -53,25 +93,43 @@ def read_examples_table_data_from_csv(example, file_path):
...
+
+                try:
+                    read_examples_table_data_from_csv(example, csv_file_path)
+                    logger.info(f"Loaded data from CSV file: {csv_file_path} for example table.")
+                except ValueError as ve:
+                    logger.warning(str(ve))
+                except FileNotFoundError as ve:
+                    logger.warning(str(ve))
+            else:
+                logger.info(f"No @from_file tag found for example table in {feature.filename}, example: {example.name}")
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:135]: DONT-CRY-IF-MISSING: That is the normal case in many situations.

File: behave/contrib/csv_table_from_file.py [hunk 3]
```diff
@@ -53,25 +93,43 @@ def read_examples_table_data_from_csv(example, file_path):
...
-
+            csv_file_path = None
+
+            # Find the @from_file={csv_filename} tag
+            for tag in example.tags:
+                if tag.startswith('@from_file='):
+                    csv_file_path = tag.split('=', 1)[1]
+                    break
+
+            if csv_file_path:
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:122]: AS ALREADY SAID: Use EARLY-BAILOUT, like

```py
if not csv_file_path:
    continue
```

File: behave/contrib/csv_table_from_file.py [hunk 4]
```diff
@@ -53,25 +93,43 @@ def read_examples_table_data_from_csv(example, file_path):
...
-            logger.warning(f"CSV file not found: {csv_file_path}")
-            continue
-
         for example in scenario_outline.examples:
-            read_examples_table_data_from_csv(example, csv_file_path)
-
+            csv_file_path = None
+
+            # Find the @from_file={csv_filename} tag
+            for tag in example.tags:
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:117]: CANDIDATE FOR HELPER-FUNCTION: `select_marker_tag_and_extract_filename(tags) -> Optional[str]`

File: behave/contrib/csv_table_from_file.py [hunk 5]
```diff
@@ -1,43 +1,83 @@
-import logging
+# -*- coding: UTF -*-
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:1]: Mark this file as python3-only (if you want to use type annotations).
HINT: 
* `behave` supports currently still Python 2.7.
* Use type annotations causes annoying problems if Python 2.7 is still used.

File: behave/contrib/csv_table_from_file.py [hunk 6]
```diff
@@ -1,43 +1,83 @@
...
+
+    Feature: Example Feature
+
+    Scenario Outline: Example Scenario
+        Given get <username> and <email> and <password>
+
+        @from_file=testdata.csv
+        Examples:
+          | username | email   | password |
+          | .        | .       | .        |
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:21]: The second table row without any values (or using dots) is not really necessary (I think).

File: behave/contrib/csv_table_from_file.py [hunk 7]
```diff
@@ -53,25 +93,43 @@ def read_examples_table_data_from_csv(example, file_path):
...
+                    break
+
+            if csv_file_path:
+                if not os.path.isabs(csv_file_path):
+                    base_path = os.path.dirname(feature.filename)
+                    csv_file_path = os.path.join(base_path, csv_file_path)
+
+                try:
+                    read_examples_table_data_from_csv(example, csv_file_path)
+                    logger.info(f"Loaded data from CSV file: {csv_file_path} for example table.")
```
Comment (jenisys) [inline behave/contrib/csv_table_from_file.py:129]: HINT: Think (and decide) if it is not better to move the logging statement before the function-call that performs the processing. Wording must be adapted in this case.

File: tests/unit/test_csv_table_from_file.py [hunk 8]
```diff
@@ -1,123 +1,89 @@
...
-    preprocess_and_read_examples_table_data_from_csv
+    preprocess_and_read_examples_table_data_from_csv, logger
+from behave.model import Table, Feature, Examples
 
 
-class Example:
-    def __init__(self):
-        self.table = None
+@pytest.fixture
+def mock_example():
```
Comment (jenisys) [inline tests/unit/test_csv_table_from_file.py:12]: Mmh, you rely too much on mocks.

* Therefore, you test something what you think is true, but the reality differs.
* It is much simpler to use the existing model classes that is used by the real code.
* See comment above `tag.startswith("@from_file=")`

DEFICIENCY:
* AFAIK by scanning over your tests: There is no test that really checks the functionality works as expected

File: tests/unit/not_csv.txt
Comment (jenisys) [inline tests/unit/not_csv.txt:1]: NOT NECESSARY: Can be created in a tmp_path by the test.

File: tests/unit/test_csv_table_from_file.py [hunk 9]
```diff
@@ -1,123 +1,89 @@
...
 
-    Args:
-        mock_exists (MagicMock): Mock for os.path.exists.
-        mock_feature (Mock): Mocked Feature object.
-        mock_scenario_outline (Mock): Mocked ScenarioOutline object.
-        caplog (pytest.LogCaptureFixture): Fixture for capturing log messages.
-    """
+    script_dir = dirname(abspath(__file__))
+    csv_filename = 'test.csv'
+    mock_examples_with_tag.tags = ['@from_file={}'.format(join(script_dir, csv_filename))]
```
Comment (jenisys) [inline tests/unit/test_csv_table_from_file.py:66]: With AI, I would say: 'Hallucinating tags ...'

**Review by jenisys — CHANGES_REQUESTED** [review]
Getting closer, but:

* Implementation will not work as expected, related to using/expecting `@tags`.

File: behave/contrib/csv_table_from_file.py (hunk 2 above)
Comment (rajanmanwansh) [reply behave/contrib/csv_table_from_file.py:135]: Resolved

File: behave/contrib/csv_table_from_file.py (hunk 3 above)
Comment (rajanmanwansh) [reply behave/contrib/csv_table_from_file.py:122]: Resolved

File: behave/contrib/csv_table_from_file.py (hunk 5 above)
Comment (rajanmanwansh) [reply behave/contrib/csv_table_from_file.py:1]: Resolved

File: behave/contrib/csv_table_from_file.py (hunk 6 above)
Comment (rajanmanwansh) [reply behave/contrib/csv_table_from_file.py:21]: Resolved

File: behave/contrib/csv_table_from_file.py (hunk 7 above)
Comment (rajanmanwansh) [reply behave/contrib/csv_table_from_file.py:129]: resolved

File: tests/unit/test_csv_table_from_file.py (hunk 8 above)
Comment (rajanmanwansh) [reply tests/unit/test_csv_table_from_file.py:12]: Resolved

File: tests/unit/not_csv.txt
Comment (rajanmanwansh) [reply tests/unit/not_csv.txt:1]: Resolved

File: tests/unit/test_csv_table_from_file.py (hunk 9 above)
Comment (rajanmanwansh) [reply tests/unit/test_csv_table_from_file.py:66]: Resolved

**Comment by rajanmanwansh** [comment]
Please review the latest change.

**Comment by jenisys** [comment]
PULL-REQUEST manually merged with modifications:

* Squash all commits into one
* See commits for details of changes/modifications
* Testsuite enhanced to cover BAD-CSV-TABLE-DATA cases, too.
* Use `csv.DictReader` now to better handle BAD-CSV-TABLE-DATA cases.
* Test to ensure that changed Examples.table causes that ScenarioOutline.scenarios are rebuild (with the new table data)

INTEGRATED INTO: behave v1.2.7.dev6

**Comment by vivek007ranjan** [comment]
@jenisys @rajanmanwansh 
Great work, this is a real time saver for implementing the extra layer. Please include this in Behave's main release and update the main documentation accordingly.

**Comment by hasinaxp** [comment]
@jenisys  @rajanmanwansh nice work....it would also work as a workaround for multiline data in column data in example table

**Comment by maheshsattala** [comment]
cool stuff...
Great work.

**Comment by ravenbirdb1b** [comment]
@jenisys @rajanmanwansh This is a valuable enhancement. Could you please add this to the main documentation regarding its usage?

**Pull request closed without merging**